diff --git a/src/gfx/fan/GfxEnv.fan b/src/gfx/fan/GfxEnv.fan
--- a/src/gfx/fan/GfxEnv.fan
+++ b/src/gfx/fan/GfxEnv.fan
@@ -85,5 +85,16 @@
   **
   abstract Int fontWidth(Font f, Str s)
 
+  abstract Graphics getImageGraphics(Image image)
+
+  abstract Image createImage(Size size)
+
+  abstract Str? getFromClipboard()
+
+  abstract Str? getFromSelectionClipboard()
+
+  abstract Void copyToClipboard(Str text)
+
+  abstract Void copyToSelectionClipboard(Str text)
 
 }
\ No newline at end of file
